### PR TITLE
Download kind binary

### LINKED ISSUE #105: [kwokctl] Download kind binary automatically if necessary
### What would you like to be added?

When using Kind as a runtime, download the kind if it doesn't exist

### Why is this needed?

Optimize the experience

### PR DESCRIPTION
**Pull request by JarHMJ**
Signed-off-by: huangminjie <minjie.huang@daocloud.io>

#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #105

#### Does this PR introduce a user-facing change?

```release-note
Download kind binary
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kwokctl/runtime/kind/cluster.go [hunk 1]
```diff
@@ -430,3 +440,25 @@ func (c *Cluster) EtcdctlInCluster(ctx context.Context, stm utils.IOStreams, arg
...
+	conf, err := c.Config()
+	if err != nil {
+		return err
+	}
+	runtime := conf.Runtime
+	if runtime == "kind" {
+		err := utils.Exec(ctx, "", utils.IOStreams{}, runtime, "version")
+		if err != nil {
+			// kind does not exist, try to download it
+			kindPath := utils.PathJoin(conf.Workdir, "bin", "kind"+vars.BinSuffix)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/cluster.go:455]: Need to return this path, which is not in the environment variable PATH

**Review by wzshiming — CHANGES_REQUESTED** [review]

**Review by wzshiming — CHANGES_REQUESTED** [review]
Remove the step

https://github.com/kubernetes-sigs/kwok/blob/26c5a1c3c70a4a6492a990cf632b8210bc8d0eba/.github/workflows/test.yaml#L130-L135

File: pkg/kwokctl/runtime/kind/cluster.go [hunk 2]
```diff
@@ -430,3 +440,25 @@ func (c *Cluster) EtcdctlInCluster(ctx context.Context, stm utils.IOStreams, arg
...
+
+// preDownloadKind pre-download and cache kind
+func (c *Cluster) preDownloadKind(ctx context.Context) (string, error) {
+	conf, err := c.Config()
+	if err != nil {
+		return "", err
+	}
+	runtime := conf.Runtime
+	if runtime == "kind" {
+		err := utils.Exec(ctx, "", utils.IOStreams{}, runtime, "version")
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/cluster.go:452]: Use `exec.LookPath` to check the existence of the command

File: pkg/kwokctl/runtime/kind/cluster.go [hunk 3]
```diff
@@ -430,3 +440,25 @@ func (c *Cluster) EtcdctlInCluster(ctx context.Context, stm utils.IOStreams, arg
...
 		append([]string{"exec", "-i", "-n", "kube-system", etcdContainerName, "--", "etcdctl", "--endpoints=127.0.0.1:2379", "--cert=/etc/kubernetes/pki/etcd/server.crt", "--key=/etc/kubernetes/pki/etcd/server.key", "--cacert=/etc/kubernetes/pki/etcd/ca.crt"}, args...)...)
 }
+
+// preDownloadKind pre-download and cache kind
+func (c *Cluster) preDownloadKind(ctx context.Context) (string, error) {
+	conf, err := c.Config()
+	if err != nil {
+		return "", err
+	}
+	runtime := conf.Runtime
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/cluster.go:450]: kind runtime has only one implementation, no need for this

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/apis/internalversion/zz_generated.conversion.go [hunk 4]
```diff
@@ -243,6 +244,7 @@ func autoConvert_v1alpha1_KwokctlConfigurationOptions_To_internalversion_Kwokctl
 	out.EtcdVersion = in.EtcdVersion
 	out.PrometheusVersion = in.PrometheusVersion
 	out.DockerComposeVersion = in.DockerComposeVersion
+	// WARNING: in.KindVersion requires manual conversion: does not exist in peer-type
```
Comment (wzshiming) [inline pkg/apis/internalversion/zz_generated.conversion.go:247]: Why is there no KindVersion

File: Makefile [hunk 5]
```diff
@@ -154,7 +154,7 @@ integration-test:
 ## e2e-test: Run e2e tests
 .PHONY: e2e-test
 e2e-test:
-	@./hack/requirements.sh kubectl buildx kind
+	@./hack/requirements.sh kubectl buildx
```
Comment (wzshiming) [inline Makefile:157]:
```suggestion
	@./hack/requirements.sh kubectl buildx kind
```

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/apis/internalversion/zz_generated.conversion.go (hunk 4 above)
Comment (JarHMJ) [reply pkg/apis/internalversion/zz_generated.conversion.go]: I found that KindVersion is unused. But I will add it

File: pkg/apis/internalversion/zz_generated.conversion.go [hunk 6]
```diff
@@ -283,6 +286,8 @@ func autoConvert_v1alpha1_KwokctlConfigurationOptions_To_internalversion_Kwokctl
 	out.PrometheusBinaryTar = in.PrometheusBinaryTar
 	// WARNING: in.DockerComposeBinaryPrefix requires manual conversion: does not exist in peer-type
 	out.DockerComposeBinary = in.DockerComposeBinary
+	// WARNING: in.KindBinaryPrefix requires manual conversion: does not exist in peer-type
```
Comment (wzshiming) [inline pkg/apis/internalversion/zz_generated.conversion.go:289]: `KindBinaryPrefix` It should be needed

**Review by wzshiming — APPROVED** [review]

File: pkg/consts/consts.go [hunk 7]
```diff
@@ -33,6 +33,9 @@ var (
 	DockerComposeVersion      = "2.13.0"
 	DockerComposeBinaryPrefix = "https://github.com/docker/compose/releases/download"
 
+	KindVersion      = "0.17.0"
+	KindBinaryPrefix = "https://github.com/kubernetes-sigs/kind/releases/download/"
```
Comment (wzshiming) [inline pkg/consts/consts.go:37]:
```suggestion
	KindBinaryPrefix = "https://github.com/kubernetes-sigs/kind/releases/download"
```

**Review by wzshiming — CHANGES_REQUESTED** [review]

**Pull request merged**